File: Data Structures & Algorithms/matrixDFS/submission-0.py
class Solution:
    def countPaths(self, grid: List[List[int]]) -> int:
        m, n = len(grid), len(grid[0])

        def dfs(r, c, visited, m, n):
            if (min(r, c) < 0 or r == m or c == n 
                or (r,c) in visited or grid[r][c] == 1):
                return 0

            if r == m - 1 and c == n - 1:
                return 1

            visited.add((r, c))

            count = 0
            count += dfs(r + 1, c, visited, m, n)
            count += dfs(r - 1, c, visited, m, n)
            count += dfs(r, c + 1, visited, m, n)
            count += dfs(r, c - 1, visited, m, n)

            # backtrack
            visited.remove((r, c))
            return count

        count = dfs(0, 0, set(), m, n)
        return count

File: Data Structures & Algorithms/count-number-of-islands/submission-0.java
class Solution {
    public int numIslands(char[][] grid) {
        int islands = 0;
        int n = grid.length;
        int m = grid[0].length;
        boolean[][] visited = new boolean[n][m];

        for(int row = 0; row < n; row++){
            for(int col = 0; col < m; col++){
                if(!visited[row][col] && grid[row][col] == '1'){
                    islands++;
                    dfs(grid, visited, row, col, n, m);
                }
            }
        }

        return islands;
    }

    private void dfs(char[][] grid, boolean[][] visited, int row, int col, int n, int m){
        visited[row][col] = true;
        int[][] directions = {{-1, 0}, {1, 0}, {0, 1}, {0, -1}};

        for(int[] direction: directions){
            int newRow = row + direction[0];
            int newCol = col + direction[1];
            if(isSafe(newRow, newCol, n, m) && !visited[newRow][newCol] && grid[newRow][newCol] == '1'){
                dfs(grid, visited, newRow, newCol, n, m);
            }
        }
    }

    private boolean isSafe(int row, int col, int n, int m){
        if(row >= 0 && row < n && col >= 0 && col < m){
            return true;
        }
        return false;
    }
}
